### DISCUSSION TITLE
TX2 BootROM and redundant boot slots

### DISCUSSION THREAD

**Discussion by graycat** [OE4T/meta-tegra#1257] [Q&A] [answered]
First of all, thanks for all the hard work you've all been doing here! It's been extremely valuble for a TX2i based product we've been developing at my company.

Speaking of which, we've recently switched over to using a cboot-only A/B redundant boot flow using RAUC with `tegra-boot-tools`, which (once I gave up on the idea of trying to keep u-boot in the chain, for now at least) was very straightforward to setup and is working great (we're using the `kirkstone-l4t-r32.7.x` branch). I've been digging through Nvidia docs and spelunking the `meta-tegra` and `tegra-boot-tools` source to make sure I fully understand the redundant boot flow, and there's one piece I'm not entirely sure on yet: 

From what I can tell the `BootROM` always looks at the `BCT` partition only (ignoring `SMD`), and it always works its way through the bootloader entries in the table sequentially. So unless the `MB1` partition gets corrupted, it's always going to jump to `MB1` and not `MB1_b` (being the second entry in the `BCT`). Then from there, the running `MB1` reads its corresponding (i.e. `*_b` if `MB1_b` is running) `MB1-BCT` for init settings and `SMD` to populate the scratch register with the slot info, then jumps to the `TBoot-BPMP` in `MB2` or `MB2_b` based on the contents of its `SMD`. In other words, the `MB1`, `MB1-BCT`, and `SMD` partitions used are always chosen by the `BootROM` based entirely off verifying `MB1` against what's stored in the `BCT` entry, and the actual A/B slot booting process starts at the `MB1->MB2` handoff.  Is this an accurate understanding? 

Oh and follow-up question: I'm sure I can find the answer with a bit more digging, but when running `tegra-bootloader-update` on the t186 to write a new bup payload to the inactive slot, does `BCT` get updated with a new table order so the new `MB1` (if it's changed and has beeen updated) is the first entry? Or will the BL entry order in `BCT` always remain the same?

**Comment by madisongh** [comment, marked as answer]
Your description of those early boot stages sounds right to me.  When updating, the BCT payload is written to each of the three BCT slots, and the mb1 payload is written to both the `mb1` and `mb1_b` partitions. The order should always be the same for the mb1 stage, as far as I know - mb1 always primary, with mb1_b only used as a backup.

**Reply by graycat** [reply]
OK great, thanks for confirming.
